src/PluginsRegistry.js

```diff
@@ -6,46 +6,59 @@
  * @api public
  */
 export default class PluginsRegistry {
-  registry: Array<Function>;
+  registry: Array<Object>
+  jss: Object
 
-  constructor(): void {
+  constructor(jss: Object): void {
     this.registry = []
+    this.jss = jss
   }
 
   /**
-   * Register plugin. Passed function will be invoked with a rule instance.
+   * Register plugin.
+   * If plugin is a function, its an onRule hook.
+   * Also we call .onSetup hook here.
    *
-   * @param {Function} fn
    * @api public
    */
-  use(fn: Function): void {
-    this.registry.push(fn)
+  use(plugin: Function | Object): void {
+    if (typeof plugin === 'function') {
+      plugin = {onRule: plugin}
+    }
+    this.registry.push(plugin)
+    if (plugin.onSetup) plugin.onSetup(this.jss)
   }
 
   /**
-   * Execute all registered plugins on all rules.
+   * Call all .onRule hooks for passed rules.
    *
-   * @param {Rule|Array} rules
    * @api public
    */
-  run(rules: Array<Object> | Object): void {
-    if (Array.isArray(rules)) {
-      rules.forEach(this.runOne, this)
-      return
-    }
+  handleRules(rules: Array<Object>): void {
+    rules.forEach(this.handleRule, this)
+  }
 
-    this.runOne(rules)
+  /**
+   * Call all .onRule hooks for passed rule.
+   *
+   * @api public
+   */
+  handleRule(rule: Object): void {
+    for (let index = 0; index < this.registry.length; index++) {
+      const plugin = this.registry[index]
+      if (plugin.onRule) plugin.onRule(rule)
```

Comment:
An alternative to conditionally checking every plugin in the registry for the presence of a specific hook every time, is to register the hooks in `.use`. This can be done by having an object holding an array for each hook, eg:

pseudo code:

```js
hooks = {
  onSheet: [],
  onRule: [],
};

use(plugin) {
  Object.keys(plugin).forEach((key) => {
    if (this.hooks[key]) {
      this.hooks[key].push(plugin[key]);
    }
  });
}

handleSheet(sheet) {
  this.hooks.onSheet.forEach((n) => n(sheet));
}
```